### PR TITLE
[SPARK-50333][SQL] Codegen Support for `CsvToStructs` (by Invoke & RuntimeReplaceable)

### PR DESCRIPTION
**Pull request by panbingkun**
### What changes were proposed in this pull request?
The pr aims to add `Codegen` Support for `CsvToStructs`(`from_csv`).

### Why are the changes needed?
- improve codegen coverage.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Pass GA & Existed UT (eg: CsvFunctionsSuite#`*from_csv*`)

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/csvExpressions.scala [hunk 1]
```diff
@@ -85,71 +86,33 @@ case class CsvToStructs(
...
 
   override def prettyName: String = "from_csv"
 
+  @transient
+  private lazy val evaluator: CsvToStructsEvaluator = CsvToStructsEvaluator(
+    options, nullableSchema, nameOfCorruptRecord, timeZoneId, requiredSchema)
+
+  override def nullSafeEval(input: Any): Any = {
+    evaluator.evaluate(input.asInstanceOf[UTF8String])
+  }
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/csvExpressions.scala:107]: Why do you need this? What happens if you don't override?

Comment (panbingkun) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/csvExpressions.scala]: Yes, you're right. 
I overlooked this and it has been updated, 
thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/csv/CsvExpressionEvalUtils.scala [hunk 2]
```diff
@@ -18,10 +18,76 @@ package org.apache.spark.sql.catalyst.expressions.csv
...
+import org.apache.spark.SparkException
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.csv.{CSVInferSchema, CSVOptions, UnivocityParser}
+import org.apache.spark.sql.catalyst.expressions.ExprUtils
+import org.apache.spark.sql.catalyst.util.{FailFastMode, FailureSafeParser, PermissiveMode}
+import org.apache.spark.sql.errors.QueryCompilationErrors
 import org.apache.spark.sql.types.{DataType, NullType, StructType}
 import org.apache.spark.unsafe.types.UTF8String
 
+case class CsvToStructsEvaluator(
```
Comment (dtenedor) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/csv/CsvExpressionEvalUtils.scala:30]: Shall we have a high-level comment for this class, explaining what it represents as an object, and who creates it and invokes it for what purpose?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/csv/CsvExpressionEvalUtils.scala [hunk 3]
```diff
@@ -18,10 +18,76 @@ package org.apache.spark.sql.catalyst.expressions.csv
...
+  // This converts parsed rows to the desired output by the given schema.
+  @transient
+  private lazy val converter = (rows: Iterator[InternalRow]) => {
+    if (rows.hasNext) {
+      val result = rows.next()
+      // CSV's parser produces one record only.
+      assert(!rows.hasNext)
+      result
+    } else {
+      throw SparkException.internalError("Expected one row from CSV parser.")
```
Comment (dtenedor) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/csv/CsvExpressionEvalUtils.scala:46]: optional: we can invert the logic and throw the exception early, de-denting the rest of the block:

```
if (!rows.hasNext) {
  throw SparkException.internalError("Expected one row from CSV parser.")
}
...
```

**Review by dtenedor — COMMENTED** [review]
Thanks for adding the codegen support!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/csv/CsvExpressionEvalUtils.scala (hunk 3 above)
Comment (panbingkun) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/csv/CsvExpressionEvalUtils.scala:44]: Yeah, I like this writing style, the logic looks `more intuitive`, and it has been updated. 
Thank you!

**Review by MaxGekk — APPROVED** [review]

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @panbingkun and @dtenedor for review.

**Pull request closed without merging**